File: src/2024/day_04/golang/main.go
package main

import (
	_ "embed"
	"strconv"
	"strings"
)

//go:embed inputs.txt
var input string

func searchDirection(text [][]string, start_x int, start_y int, dir int, matchLen int, matchStr string) (bool, []int) {
	var next []int
	switch dir {
	case 1:
		next = []int{-1, -1}
	case 2:
		next = []int{0, -1}
	case 3:
		next = []int{1, -1}
	case 4:
		next = []int{1, 0}
	case 5:
		next = []int{1, 1}
	case 6:
		next = []int{0, 1}
	case 7:
		next = []int{-1, 1}
	case 8:
		next = []int{-1, 0}
	}

	max_x := start_x + matchLen*next[0]
	max_y := start_y + matchLen*next[1]

	if max_y < 0 || max_x < 0 || max_y >= len(text) || max_x >= len(text[start_y]) {
		return false, []int{-1, -1}
	}

	for idx := range matchLen {
		if text[start_y+next[1]*(idx+1)][start_x+next[0]*(idx+1)] != strings.Split(matchStr, "")[idx] {
			return false, []int{-1, -1}
		}
	}

	return true, []int{start_x + next[0]*(matchLen-1), start_y + next[1]*(matchLen-1)}
}

func mainPart1(text [][]string) {
	count := 0
	for idx_y, line := range text {
		for idx_x, char := range line {
			if char == "X" {
				for _, dir := range []int{1, 2, 3, 4, 5, 6, 7, 8} {
					match, _ := searchDirection(text, idx_x, idx_y, dir, 3, "MAS")
					if match {
						count++
					}
				}
			}
		}
	}
	println("XMAS appears", strconv.Itoa(count), "times")
}

type coords struct {
	x, y int
}

func mainPart2(text [][]string) {
	aLocations := map[coords]bool{}
	count := 0

	for idx_y, line := range text {
		for idx_x, char := range line {
			if char == "M" {
				for _, dir := range []int{1, 3, 5, 7} {
					match, m_coords := searchDirection(text, idx_x, idx_y, dir, 2, "AS")
					if match {
						coord := coords{m_coords[0], m_coords[1]}
						if aLocations[coord] {
							count++
						} else {
							aLocations[coord] = true
						}
					}
				}
			}
		}
	}
	println("X-MAS appears", strconv.Itoa(count), "times")
}

func main() {
	wordSearchLines := strings.Split(input, "\n")
	wordSearch := make([][]string, len(wordSearchLines))
	for idx, line := range wordSearchLines {
		wordSearch[idx] = strings.Split(line, "")
	}
	mainPart1(wordSearch)
	mainPart2(wordSearch)
}

File: src/2024/day_04/python/main.py

# Directions:
# 1  2  3
# 8  X  4
# 7  6  5

def search_direction(letters:list[list[str]], start_x:int, start_y:int, dir:int, match_len:int, match_str:str) -> (bool, (int, int)):
    match dir:
        case 1:
            next = [(-1, -1), (-2, -2), (-3, -3)]
        case 2:
            next = [(0, -1), (0, -2), (0, -3)]
        case 3:
            next = [(1, -1), (2, -2), (3, -3)]
        case 4:
            next = [(1, 0), (2, 0), (3, 0)]
        case 5:
            next = [(1, 1), (2, 2), (3, 3)]
        case 6:
            next = [(0, 1), (0, 2), (0, 3)]
        case 7:
            next = [(-1, 1), (-2, 2), (-3, 3)]
        case 8:
            next = [(-1, 0), (-2, 0), (-3, 0)]

    max_x = start_x + next[match_len-1][0]
    max_y = start_y + next[match_len-1][1]

    if max_y < 0 or max_x < 0 or max_y >= len(letters) or max_x >= len(letters[start_y]):
        return False, (0, 0)
    

    val = match_str == ''.join([letters[start_y + next[i][1]][start_x + next[i][0]] for i in range(match_len)])

    return val, (start_x + next[match_len-2][0], start_y + next[match_len-2][1])

def main_part1(word_search: list[list[str]]):
    count = 0
    for idx_y, line in enumerate(word_search):
        for idx_x, letter in enumerate(line):
            if letter == 'X':
                count += sum([search_direction(word_search, idx_x, idx_y, dir, 3, 'MAS')[0] for dir in range(1,9)])
    
    print(f"XMAS appears {count} times")

def main_part2(word_search: list[list[str]]):
    
    x_mas_matches = []
    for idx_y, line in enumerate(word_search):
        for idx_x, letter in enumerate(line):
            if letter == 'M':
                x_mas_matches.extend(filter(lambda x: x[0], [search_direction(word_search, idx_x, idx_y, dir, 2, 'AS') for dir in [1, 3, 7, 5]]))

    x_mas_set = set(x_mas_matches)
    count = len(x_mas_matches) - len(x_mas_set)

    print(f"X-MAS appears {count} times")

if __name__ == "__main__":
    with open("../inputs.txt") as file:
        mult_string = file.readlines()
    word_search = [list(line.strip()) for line in mult_string]

    main_part1(word_search)
    main_part2(word_search)
    